library/Icingadb/Common/Auth.php

```diff
@@ -169,19 +169,28 @@ public function applyRestrictions(Query $query)
                     }
 
                     if ($applyHostRestriction && ($restriction = $role->getRestrictions('icingadb/filter/hosts'))) {
-                        $roleFilter->add($this->parseRestriction($restriction, 'icingadb/filter/hosts'));
+                        $hostFilter = $this->parseRestriction($restriction, 'icingadb/filter/hosts');
+                        // TODO: This is really a very poor solution is therefore only a quick fix.
+                        //  We need to somehow manage to make this more enjoyable and creative!
+                        if ($orgQuery instanceof UnionQuery && strpos($restriction, 'hostgroup.name') !== false) {
```

Comment:
Eeeh, no. This introduces another bug: What if the restriction is `host.name=foo|hostgroup.name=bar`? Then the `host.name=foo` filter is processed as subquery. :boom: 

You need to traverse through the filter object, I'm afraid.

Reply:
Or wait, no, `hostgroup.name` is processed as one, which you want to avoid, right? :thinking: Anyway, forcing optimization should be done on the individual rules, not the entire chain.